engine/src/main/java/org/terasology/engine/utilities/procedural/PerlinNoise.java

```diff
@@ -72,13 +72,22 @@ public PerlinNoise(long seed, int gridDim) {
      */
     @Override
     public float noise(float posX, float posY, float posZ) {
-        int xInt = Math.floorMod(TeraMath.floorToInt(posX), permCount);
-        int yInt = Math.floorMod(TeraMath.floorToInt(posY), permCount);
-        int zInt = Math.floorMod(TeraMath.floorToInt(posZ), permCount);
 
-        float x = posX - TeraMath.fastFloor(posX);
-        float y = posY - TeraMath.fastFloor(posY);
-        float z = posZ - TeraMath.fastFloor(posZ);
+        // Domain rotation removes Perlin's characteristic square artifacts from the XZ planes, by pointing Y up the grid's main diagonal.
```

Comment:
I still don't really understand what rotation is actually being performed here, and it's not clear what exactly the math below is doing either. I'd appreciate more explanation - for example, a comment here saying how much it's rotating around what axis, and comments below on things like the magic numbers explaining where you got them.

Reply:
> It's also my impression that there's no reason to use Perlin noise anymore, and that Simplex doesn't have these artifacts in the first place. It's probably better to switch existing uses of Perlin noise to Simplex than to improve the Perlin noise. If there's actually some reason for something to continue using Perlin noise, then we should un-deprecate it and this change makes sense, but I don't think that's the case.

I agree the noise isn't great in its unmitigated/unrotated form, and would agree with decisions to phase it out. With the rotation, though, I've founds the results to be quite nice when used for overhang-enabled voxel terrain. I use a variant of it in [this project](https://github.com/KdotJPG/Simply-Improved-Terrain). If including the change and removing the deprecation, I could also rename the noise to "RPerlin" or "RotoPerlin" to distinguish it from the unmitigated algorithm, if that would make sense for this project. Any user-visible references could be updated accordingly to go with that (e.g. the "Perlin" world type). I recently began using this name [here](https://github.com/KdotJPG/Noise-Extras/blob/master/CachedColumnNoise/RotoperlinCachedColumnNoise.java) to refer to implementations of the old Perlin algorithm that have the built-in rotational improvement.

Re: Simplex, there are some artifacts present in the simplex noise given the gradient tables it uses, but they're not of the same strongly grid-aligned nature. I have some OpenSimplex2 code that I'd like to submit that would improve them, I just wound up tackling this first.

> Rotation

I explained a lot of it in [this other comment](https://github.com/MovingBlocks/Terasology/pull/4802#discussion_r660157980). Let me know if there's anything else I can clarify. Happy to add more details to a comment in the actual as well! The rotational axis wouldn't really describe much about what the rotation does, but perhaps I could mention that the rotation rotates the Y vector 45 degrees to point down all three of X, Y, Z equally.